### DISCUSSION TITLE
how does docker-build determines tty / plain?

### DISCUSSION THREAD

**Discussion by kasperk81** [moby/moby#47755] [Q&A]
how does docker-build determine it's not tty?

in github actions ci, `docker build .` automatically switch to text mode, while on appveyor ci it thinks tty is available which makes the output very hard to read (it keeps repeating the whole dockerfile with every line printed during progress, so imagine 60K lines surprise when running `apt install`...). to avoid that, every docker-build line needs `--progress=plain`.

but what _exactly_ makes docker decide this? setting TERM=xterm-256color, unsetting TERM or setting it to `dumb`  on both CIs as well as piping both to tee didn't make any difference

**Comment by thaJeztah** [comment]
It detects if there's a TTY attached on STDERR (using `console.ConsoleFromFile()`);
https://github.com/docker/buildx/blob/e344e2251bbfebf3c60072abe249dd76e4dac0a2/commands/build.go#L310-L362

Which is a utility in containerd's `console` module;
https://github.com/containerd/console/blob/8f6c4e4faef5a326d2cd907097d04c0239ee5e2f/console.go#L83-L90

Actual detection is platform-dependent, here's the handling for Unix and Windows;

https://github.com/containerd/console/blob/8f6c4e4faef5a326d2cd907097d04c0239ee5e2f/console_unix.go#L124-L130
https://github.com/containerd/console/blob/8f6c4e4faef5a326d2cd907097d04c0239ee5e2f/console_windows.go#L204-L210


It's likely that CI in your case as a TTY attached, in which case the default (`auto`) will print output for a console / TTY. Instead of updating all your build build commands, you may consider setting the `BUILDKIT_PROGRESS` environment-variable instead; that allows you to set the same options as is used in the `--progress` flag; it's documented in the Environment Variables section in the documentation; https://docs.docker.com/engine/reference/commandline/cli/#environment-variables

**Comment by kasperk81** [comment]
got it, thanks! so `docker build ... | tee` doesn't switch to plain text, but `docker build ... 2>&1 | tee` does. most software i've seen which detect tty to select output format (plain vs. interactive) make this decision based on whether or not stdout is redirected. docker is doing a unique thing this way. aligning with that would be nice `console.ConsoleFromFile(os.Stderr)` -> `console.ConsoleFromFile(os.Stdout)`

> `BUILDKIT_PROGRESS`

will send a pr to appveyor 👍
